<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="UTF-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Document</title>
 </head>
 <body>
  <script>
   function cuadrado_promise(value, callback) {
    if (typeof value !== "number") {
     return Promise.reject(
      `Error, el valor "${value}" ingresado no es un nÃºmero`
     );
    }
    return new Promise((resolve, reject) => {
     setTimeout(() => {
      resolve({
       value,
       result: value * value,
      });
     }, 0 | (Math.random() * 1000));
    });
   }

   async function funcion_asincrona_declarada() {
    try {
     console.log("Inicio Async Function");

     let obj = await cuadrado_promise(0);
     console.log(`Async Function: ${obj.value}, ${obj.result}`);

     obj = await cuadrado_promise(1);
     console.log(`Async Function: ${obj.value}, ${obj.result}`);

     obj = await cuadrado_promise(2);
     console.log(`Async Function: ${obj.value}, ${obj.result}`);

     obj = await cuadrado_promise(3);
     console.log(`Async Function: ${obj.value}, ${obj.result}`);

     obj = await cuadrado_promise(4);
     console.log(`Async Function: ${obj.value}, ${obj.result}`);

     obj = await cuadrado_promise(5);
     console.log(`Async Function: ${obj.value}, ${obj.result}`);

					console.log("Final de Async Function");
    } catch (err) {
					console.error(err)
				}
   }
   funcion_asincrona_declarada();
  </script>
 </body>
</html>
